### PR TITLE
Fix duplicate accessibilityLabels & app crash on iOS

### LINKED ISSUE #21830: Wrong translation from react-native's accessibilityLabel to iOS name/label
## Environment
  OS: macOS High Sierra 10.13.6
  Node: 10.8.0
  Yarn: Not Found
  npm: 6.4.1
  Watchman: Not Found
  Xcode: Xcode 10.0 Build version 10A255
  Android Studio: 3.1 AI-173.4907809

  Packages: (wanted => installed)
    react: ^16.4.1 => 16.4.1
    react-native: ^0.55.4 => 0.55.4

## Description
When accessibilityLabel set to the element in react-native code it translates to the iOS name/label, but all parent element's name/label accumulate accessibilityLabel value of a child element.

In Appium inspector it's looks like:
```
<e1 name ='label 2 label 1'>
  <e2 name ='label 2'>
    <e3 name ='label 1'>
      <e4 name ='label 1'>
```

But in native code:
```
<e1>
  <e2 accessibilityLabel ='label 2'>
    <e3>
      <e4 accessibilityLabel ='label 1'>
```

It completely ruins all efforts to automate testing of iOS app built from react-native code.
I investigated this issue and found it appears from work of RCTRecursiveAccessibilityLabel() function, which called in RCTView.m file in this code:
```
- (NSString *)accessibilityLabel
{
  NSString *label = super.accessibilityLabel;
  if (label) {
    return label;
  }
  return RCTRecursiveAccessibilityLabel(self);
}
```
Change `return RCTRecursiveAccessibilityLabel(self);` to `return nil;` - fix this issue, so I suggest this code changes to be implemented.

### PR DESCRIPTION
**Pull request by nfriedly**
## Summary

I came across this because one of my customer's apps was freezing and crashing when my code tries to read the `accessibilityLabel` of a view containing a chat component. Because the labels are recursively generated from the text content, any sufficiently long chat ends up with all of the parent views having a ridiculously long `accessibilityLabel`. 

After 10 seconds of non-responsiveness from trying to generate all of this on the UI thread (when backgrounded), iOS decides the app has crashed and force-closes it.

Before this change, every single view in a RN app returned true for the conditional. 

I believe this change will also improve both the accessibility and the testability of RN apps, in addition to fixing that crash.

This change also allows for correct behavior with this code:

```objective-c
if ([view respondsToSelector:@selector(accessibilityLabel)])
    return view.accessibilityLabel;
```

Before this change, any RCTView could potentially scan its entire tree and return `nil`. 
After, only RCTViews that actually have an `accessibilityLabel` set will return one, and ones that don't will be skipped by the above code.

This PR supersedes or fixes #21830, #25963, #24113, #24118, https://github.com/appium/appium/issues/10654, possibly #25220, and probably a few other tickets I haven't identified.

The changes in this PR are for `RCTView`. It's worth mentioning that there is very similar code in [`RCTViewComponentView`](https://github.com/facebook/react-native/blob/ebf2fb7868d92661772ec59d6592f14786e9673b/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm#L526-L552) - I haven't dug into that as much, but I suspect it should probably get the same treatment. If desired, I can include it in this PR.

I also haven't yet reviewed the Android side of things to see what happens there, but whatever the case, the same component structure doesn't seem to be leading to crashes in Android.

I might need to change the documentation too. 

## Changelog

[iOS] [Removed] - Deleted RCTView's accessibilityLabel wrapper and RCTRecursiveAccessibilityLabel, allowing for default native behavior

## Test Plan

This is where I need a little help. What's a good way to test this? 

I spent a while digging through the RN iOS tests, I can run the tests, and even step through the harness, but I still don't feel like I fully understand what's happening. I don't have a ton of iOS or RN experience, so it may be something obvious, but I'm just not sure what I'm missing.

### CODE REVIEW

File: React/Views/RCTView.m [hunk 1]
```diff
@@ -223,7 +205,7 @@ - (NSString *)accessibilityLabel
   if (label) {
     return label;
   }
-  return RCTRecursiveAccessibilityLabel(self);
+  return nil;
 }
```
Comment (nfriedly) [inline React/Views/RCTView.m:209]: Come to think of it, I might be able to just remove this method entirely since all it's doing is calling the super now.

Comment (nfriedly) [reply React/Views/RCTView.m]: Update: removed. This should actually make it behave better with code like this (which is what my code actually does):

```objective-c
if ([view respondsToSelector:@selector(accessibilityLabel)])
    return view.accessibilityLabel;
```

**Comment by shergin** [comment]
I think @rigdern has a context why those functions are needed.

**Comment by rigdern** [comment]
@shergin Unfortunately, I don't have context on those functions. They've been in `RCTView` since the first open source commit of React Native. If we can find somebody that is intimately familiar with React Native's accessibility implementation to help us, that would be great. If not, since these functions have been around for so long, here are some things I would check before making the change to ensure it doesn't cause a regression:
  - Does `RCTView` exhibit the same issues? Or did `RCTViewComponentView` introduce them or exacerbate them? For example, perhaps in `RCTViewComponentView` hierarchies some views are marked as accessible that aren't in `RCTView` hierarchies that cause heavier use of the `accessibiiltyLabel` override.
  - Check with some app developers that have been using React Native's accessibility framework for a long time. Have they encountered any problems with React Native's `accessibilityLabel` implementation? Is it doing what they want? Would they prefer the new implementation?
  - Does Android exhibit a similar behavior and do we want to change it? When I was actively contributing to React Native, I was advocating for consistency between iOS and Android for the accessibility APIs. That way, a developer can write some accessibility code and expect it'll do the right thing on iOS and Android. However, that was a couple of years ago so I'm not familiar with how the accessibility APIs have evolved or the current philosophy behind them.

**Comment by nfriedly** [comment]
I spent some time testing this in the rn-tester app on a real iPhone (it took me a while to realize that VoiceOver just doesn't work in the simulator), and I have to admit that it seems to make a somewhat worse experience there. A lot of "Possible text" statements prefixing things where it would previously just read the text. And occasionally some text is omitted entirely. So this isn't the accessibility improvement I had hoped it would be.

My main goal here is to prevent the crash, which obviously impairs accessibility. I think that can be done without a regression in functionality. Here's a couple of ideas I thought of:

* Only generate `accessibilityLabels` on selectable elements, otherwise return `nil`. 
* Put a limit on the string-length, or depth, or absolute number of child nodes that `RCTRecursiveAccessibilityLabel` looks at
* Cache the results of `RCTRecursiveAccessibilityLabel` - this would prevent the exponential as `RCTRecursiveAccessibilityLabel` when my code scans every view and `RCTRecursiveAccessibilityLabel` in turn scans every sub-view. As usual, cache invalidation is the hard part there.

What do you guys think?

**Comment by nfriedly** [comment]
A couple of quick updates: 

1) We ended up working around this issue by special-casing React Native views in our code - for `RCTView`s and `RCTViewComponentView`s and their subclasses, we now use the native `accessibilityLabel` implementation from the UIView class. 

2) I have an idea for how to make this better for everyone else without negatively impacting actual accessibility: I believe that the `accessibilityLabel`s are only used by VoiceOver when `accessibilityElement` is `true` - so when `RCTView.accessibiltyLabel()` is called, I'd like to make it check that and only do the recursive string building in that case. 

I feel like there might be a better way to build the string also - maybe pass in the NSMutableString instead of creating a new one for each view. 

I'll experiment with the above when I have time, but I'm not sure when that will be now that the immediate pressure is off.

**Comment by nfriedly** [comment]
I'm going to close this PR, as there are other PRs up now that address the problem better than this one (e.g. https://github.com/facebook/react-native/pull/31924 improves performance _without_ changing behavior, https://github.com/facebook/react-native/pull/31924 improves the performance _and_ the behavior.)

**Pull request closed without merging**